build_test/lib/src/in_memory_reader.dart

```diff
@@ -13,57 +14,35 @@ abstract class RecordingAssetReader implements AssetReader {
   Set<AssetId> get assetsRead;
 }
 
-/// An implementation of [AssetWriter] that records outputs to [assets].
-abstract class RecordingAssetWriter implements AssetWriter {
-  Map<AssetId, List<int>> get assets;
-}
-
 /// An implementation of [AssetReader] and [AssetWriter] with primed in-memory
 /// assets.
+///
+/// TODO(davidmorgan): merge into `FileBasedReader` and `FileBasedWriter`.
```

Comment:
Why? Memory based and file based seems at odds.

Reply:
I think it will end up as one class, with a different `Filesystem` implementation for in-memory and file based.

Once all the optional functionality is refactored from inheritance hierarchy to members like `AssetFinder`, it should be possible to get rid of `AssetReaderState`, get rid of delegation, and have one concrete `AssetReaderWriter` with all the interesting/optional functionality provided via those members.

The key thing allowing this refactoring is that the inheritance hierarchy / delegates are always built so they run in the same order--so for example caching always happens underneath per-generator visibility restriction. So actually that can just be "do I have visibility restriction, if so apply it, do I have a cache, if so use it" and so on :)